### DISCUSSION TITLE
The initial structure

### DISCUSSION THREAD

**Discussion by johnhooks** [wp-blocks/wp-blocks#2] [General]
The first few packages are:

- `create-block` - just a copy of `@wordpress/create-block` right now, intent is for it to be a generalized block builder. First goal is to convert the original to TS.
- `create-block-ts-template` - just a copy of `@wordpress/create-block-tutorial-template` right now, intent is for this to be the first block template. First goal ist to convert it to TS and remove mustache.
- `scripts` - empty, but I think it could be useful

**Comment by johnhooks** [comment]
@erikyo I set it up so we have to do pull requests to the `wp-block/wp-block` repo before merging. I don't think either of us have to have a review to do so, but there are a lot of settings, we'll figure it out as we go. Also working from our own forks and doing PRs to the organization repo is probably the best way to commit changes.

**Reply by erikyo** [reply]
Perfect! it sounds like the best way to proceed, and also quite similar to how gutenberg repo works, isn't?

**Reply by johnhooks** [reply]
😉 Yes! I pulled a lot strait out of Gutenberg and TSified it.

**Reply by johnhooks** [reply]
The best part of the monorepo structure is that the build tools aren't repeated for every package. Though I would argue Gutenberg has taken it a bit far, it has a lot of packages and it makes learning the structure pretty overwhelming.

**Comment by erikyo** [comment]
As far as scripts are concerned, we can fork wp-scripts, as this way we can get customised and perhaps more advanced scripts. In the past I have had to modify some of them, for example the script-zip lacks the option to provide a custom name or path to the emitted zip. 

In addition i've created some modules that could be useful like [this one ](https://github.com/erikyo/modul-r/blob/WIP/fse/bin/imagemin.mjs) that is used to encode png images into avif, and compress the other image formats

**Reply by johnhooks** [reply]
Sounds good to me, do you think inside this repo, or on its own?

**Reply by erikyo** [reply]
maybe better inside the organization as a standalone repository?

**Reply by erikyo** [reply]
ok seems legit! But this is something that would be interesting to extend because sincerely I really like wp-scripts but sometimes I would have liked a bit more

**Reply by johnhooks** [reply]
It'll be interested how much we can change, keep it working, and improve it.

**Reply by erikyo** [reply]
mmh various ideas about how we can do this, but maybe none of this is a real solution intended that it is not 100 per cent clean. 

Anyway we could:
- Install wp-script package as a dependency (or a git submodule) and then import functions and variables into a copy of the original one.
- Fork (or maybe mirror) the wp-scripts repo. It is not very up-to-date and it would be enough to rebase the customisations in top to master origin to easily return if it is updated
- Install wp-script package as a dev-dependency and use a mix of original and custom ones, combining everything into a new scripts folder using CopyWebpackPlugin to bundle the final folder

I think having a custom wp-script package is also advantageous because then we avoid creating custom configurations for builds e.g. we could modify the wp-start/build script to be better compatible with the ts / generate types etc

**Reply by johnhooks** [reply]
All good ideas, though I agree they all have their pitfalls. What do you think of just using a fork of `gutenberg`? We could keep everything in sync with the main repo except the `create-block` and `scripts` packages. Then it would actually be easy to use the parts we like, keep things in sync with the main repo, and also have the ability to easily commit it back if it's something the community likes.

**Reply by johnhooks** [reply]
We can publish them under the organization scope and easily use them.

**Reply by johnhooks** [reply]
I did it, I still think a clean implementation of what we think a good block TypeScript project structure looks like would be good to maintain, though maybe we can eventually use the templates to create the reference example block package.

https://github.com/wp-blocks/gutenberg

**Reply by johnhooks** [reply]
Gutenberg is huge! I've never build a package that large before. It might be too cumbersome to do it directly In Gutenberg.

**Reply by erikyo** [reply]
yes I agree... working directly there is quite complex I have always preferred to clone the sub-modules and work on those. And I can confirm that doing the build is really long but it takes me about 10-15 minutes (on wsl that a bit slower than win but in this case i need node <18). 
In the end I realised that for working with node the hardware is crucial and I made myself a machine for that purpose
